// posthog-js@1.235.6/lib/src/utils/type-utils.js
import { window } from './globals';
import { knownUnsafeEditableEvent } from '../types';
import { includes } from './string-utils';
// eslint-disable-next-line posthog-js/no-direct-array-check
var nativeIsArray = Array.isArray;
var ObjProto = Object.prototype;
export var hasOwnProperty = ObjProto.hasOwnProperty;
var toString = ObjProto.toString;
export var isArray = nativeIsArray ||
    function (obj) {
        return toString.call(obj) === '[object Array]';
    };
// from a comment on http://dbj.org/dbj/?p=286
// fails on only one very rare and deliberate custom object:
// let bomb = { toString : undefined, valueOf: function(o) { return "function BOMBA!"; }};
export var isFunction = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-function-check
    return typeof x === 'function';
};
export var isNativeFunction = function (x) {
    return isFunction(x) && x.toString().indexOf('[native code]') !== -1;
};
// When angular patches functions they pass the above `isNativeFunction` check (at least the MutationObserver)
export var isAngularZonePresent = function () {
    return !!window.Zone;
};
// Underscore Addons
export var isObject = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-object-check
    return x === Object(x) && !isArray(x);
};
export var isEmptyObject = function (x) {
    if (isObject(x)) {
        for (var key in x) {
            if (hasOwnProperty.call(x, key)) {
                return false;
            }
        }
        return true;
    }
    return false;
};
export var isUndefined = function (x) { return x === void 0; };
export var isString = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-string-check
    return toString.call(x) == '[object String]';
};
export var isEmptyString = function (x) { return isString(x) && x.trim().length === 0; };
export var isNull = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-null-check
    return x === null;
};
/*
    sometimes you want to check if something is null or undefined
    that's what this is for
 */
export var isNullish = function (x) { return isUndefined(x) || isNull(x); };
export var isNumber = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-number-check
    return toString.call(x) == '[object Number]';
};
export var isBoolean = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-boolean-check
    return toString.call(x) === '[object Boolean]';
};
export var isDocument = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-document-check
    return x instanceof Document;
};
export var isFormData = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-form-data-check
    return x instanceof FormData;
};
export var isFile = function (x) {
    // eslint-disable-next-line posthog-js/no-direct-file-check
    return x instanceof File;
};
export var isError = function (x) {
    return x instanceof Error;
};
export var isKnownUnsafeEditableEvent = function (x) {
    return includes(knownUnsafeEditableEvent, x);
};
