// posthog-js@1.235.6/lib/src/extensions/exception-autocapture/type-checking.js
// Portions of this file are derived from getsentry/sentry-javascript by Software, Inc. dba Sentry
// Licensed under the MIT License
import { isFunction, isNull, isObject, isUndefined } from '../../utils/type-utils';
export function isEvent(candidate) {
    return !isUndefined(Event) && isInstanceOf(candidate, Event);
}
export function isPlainObject(candidate) {
    return isBuiltin(candidate, 'Object');
}
export function isInstanceOf(candidate, base) {
    try {
        return candidate instanceof base;
    }
    catch (_a) {
        return false;
    }
}
export function isPrimitive(candidate) {
    return isNull(candidate) || (!isObject(candidate) && !isFunction(candidate));
}
export function isError(candidate) {
    switch (Object.prototype.toString.call(candidate)) {
        case '[object Error]':
        case '[object Exception]':
        case '[object DOMException]':
        case '[object DOMError]':
            return true;
        default:
            return isInstanceOf(candidate, Error);
    }
}
export function isErrorEvent(event) {
    return isBuiltin(event, 'ErrorEvent');
}
export function isErrorWithStack(candidate) {
    return 'stack' in candidate;
}
export function isBuiltin(candidate, className) {
    return Object.prototype.toString.call(candidate) === "[object ".concat(className, "]");
}
export function isDOMException(candidate) {
    return isBuiltin(candidate, 'DOMException');
}
export function isDOMError(candidate) {
    return isBuiltin(candidate, 'DOMError');
}
